### PR TITLE
Initial version of parallelization framework

### PR DESCRIPTION
**Pull request by chenccfb**
Summary:
There are two parallelization mode: data and model parallelization.

This diff only implements data parallelization for FC node, i.e., torch.linear.

I will have a follow up diff to implement parallelization for other ops.

## Model Parallelization:

        x   param                       x   param                           x   param
         \ /                             \ /                                  \ /
         add    weight  bias     ====>    add   weight_1  bias_1    ...        add(reused)   weight_4  bias_4
           \      |      /                 \       |       /                     \             |       /
            \     |     /                   \      |      /                       \            /      /
                linear                           linear1            ...                linear4
                  |                                   \             |                   /
                clamp                                  \            |                  /
                                                                concat_linear
                                                                    |
                                                                  clamp
                                                                    |
                                                                  output

## Data Parallelization:

      x   param                           x   param
         \ /                                  \ /
         add                                  add
          |                 ==>         /    /   |     \
        linear                         /    |    |      \
          |                      slice1 slice2 slice3 slice4
         clamp                        |     |    |        |
          |                     linear1  linear2 linear3 linear4
        output                         \    |     |       /
                                         \  |     |      /
                                          concat_linear
                                               |
                                             clamp
                                               |
                                             output

Test Plan:
## Unit Tests
buck test glow/fb/fx/fx_glow:test_fx_parallelize

Differential Revision: D26812645

### CODE REVIEW

File: torch/fx/node.py [hunk 1]
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
 
         return False
 
+    def replace_input_with(self, old_input: 'Node', new_input: 'Noe'):
```
Comment (jamesr66a) [inline torch/fx/node.py:442]: s/'Noe'/'Node'/

File: torch/fx/node.py [hunk 2]
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
...
+        """
+        Replace input node with new input node.
+        """
```
Comment (jamesr66a) [inline torch/fx/node.py:445]: Please format docstrings according to the [google docstring style guide](https://google.github.io/styleguide/pyguide.html), including a description, `Args`, `Return`, and ideally an example invocation

File: torch/fx/node.py [hunk 3]
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
...
+        def maybe_replace_node(n : Node) -> Node:
+            return new_input if n == old_input else n
+
+        new_args = map_arg(self.args, maybe_replace_node)
+        new_kwargs = map_arg(self.kwargs, maybe_replace_node)
+        assert isinstance(new_args, tuple)
+        assert isinstance(new_kwargs, dict)
+        self.__update_args_kwargs(new_args, new_kwargs)
```
Comment (jamesr66a) [inline torch/fx/node.py:453]: Overall I'm fine with having this method, however I'd like to understand the UX in terms of discoverability of the `map_arg` pattern. How hard is it to know to use `map_arg`? Do you think that should be a method on `Node` that applies to both `args` and `kwargs`? (essentially a generalization of this function)

File: torch/fx/node.py (hunk 1 above)
Comment (chenccfb) [reply torch/fx/node.py:442]: Good catch!

File: torch/fx/node.py (hunk 3 above)
Comment (chenccfb) [reply torch/fx/node.py:453]: > How hard is it to know to use map_arg

I think ```map_arg``` is pretty easy to use and powerful and I am happy with that :)

**Pull request closed without merging**